examples/seismic/viscoacoustic/operators.py

```diff
@@ -63,9 +98,87 @@ def blanch_symes(model, geometry, v, p, **kwargs):
     return [u_v, u_r, u_p]
 
 
-def ren(model, geometry, v, p, **kwargs):
+def sls_2nd_order(model, geometry, v, p, **kwargs):
+    """
+    Implementation of the 2nd order viscoacoustic wave-equation from Bai (2014).
+
+    https://library.seg.org/doi/10.1190/geo2013-0030.1
+
+    Parameters
+    ----------
+    v : VectorTimeFunction
+        Particle velocity.
+    p : TimeFunction
+        Pressure field.
     """
-    Stencil created from Ren et al. (2014) viscoacoustic wave equation.
+    forward = kwargs.get('forward', True)
+    space_order = p.space_order
+    s = model.grid.stepping_dim.spacing
+    b = model.b
+    vp = model.vp
+    damp = model.damp
+    qp = model.qp
+    f0 = geometry._f0
+
+    # The stress relaxation parameter
+    t_s = (sp.sqrt(1.+1./qp**2)-1./qp)/f0
+
+    # The strain relaxation parameter
+    t_ep = 1./(f0**2*t_s)
+
+    # The relaxation time
+    tt = (t_ep/t_s)-1.
+
+    # Density
+    rho = 1. / b
+
+    r = TimeFunction(name="r", grid=model.grid, time_order=2, space_order=space_order,
+                     staggered=NODE)
+
+    if forward:
+
+        u_v = Eq(v, b * grad(p))
+
+        pde_r = r + s * (tt / t_s) * rho * div(v) - s * (1. / t_s) * r
+        u_r = Eq(r.forward, damp * pde_r)
+
+        pde_p = 2. * p - damp * p.backward + s * s * vp * vp * (1. + tt) * rho * \
+            div(v) - s * s * vp * vp * r.forward
+        u_p = Eq(p.forward, damp * pde_p)
+
+        return [u_v, u_r, u_p]
+
+    else:
+
+        # Auxiliary functions
+        w = VectorTimeFunction(name="w", grid=model.grid, time_order=2,
+                               space_order=space_order)
+        h = TimeFunction(name="h", grid=model.grid, time_order=2, space_order=space_order,
+                         staggered=NODE)
+        g = TimeFunction(name="g", grid=model.grid, time_order=2, space_order=space_order,
```

Comment:
Since these are local tmps, you may only need `VectorFunction, Function` and not time dependent, same other places

Reply:
Hi @mloubout, we understood your review and we tried this approach. When we replace `TimeFunction` by `Function` we are experiencing seg fault with the operator 😭.

Reply:
That's quite odd :O